builder/build.go

```diff
@@ -508,6 +515,37 @@ func (b *Builder) BuildImage(image *Image) error {
 			}
 		}
 
+		// if building protocols, add .proto inputs and build using
+		// 'protocjs' into /mnt/out/proto
+		if len(l.ProtoBuildJs) > 0 {
+			// add the build commands in a predictable order so
+			// the generated layer ID is deterministic
+			dirs := make([]string, 0, len(l.ProtoBuildJs))
+			for dir := range l.ProtoBuildJs {
+				dirs = append(dirs, dir)
+			}
+			sort.Strings(dirs)
+			for _, dir := range dirs {
+				paths, err := filepath.Glob(filepath.Join(dir, "*.proto"))
+				if err != nil {
+					return err
+				}
+				inputs = append(inputs, paths...)
+				outDir := filepath.Join("/mnt/out/proto", l.ProtoBuildJs[dir])
+				run = append(run,
+					fmt.Sprintf("mkdir -p %s", outDir),
+					fmt.Sprintf(`protoc -I /usr/local/include -I %s \
+						--plugin="protoc-gen-ts=/usr/local/bin/protoc-gen-ts"\
+						--js_out="import_style=commonjs,binary:%s" \
+						--ts_out="service=grpc-web:%s" \
+						/usr/local/include/google/api/annotations.proto \
+						/usr/local/include/google/api/http.proto \
+						%s
+`, dir, outDir, outDir, strings.Join(paths, " ")),
+				)
```

Comment:
@lmars I think it would need to be more like

```
protobuild: [
  {
    src: "input/dir".
    dst: "output/dir",
    lang: "js"
  }
]
```

as the js/ts build (i.e. the dashboard) will have a different output directory than input (unlike the go build). I don't think we need a list of langs as it seems unlikely there would be multiple being build to the same output dir (and js/ts should always be part of the same build). WDYT?

Reply:
SGTM 👍